src/coreclr/ToolBox/superpmi/superpmi-shared/compileresult.cpp

```diff
@@ -742,81 +742,130 @@ void CompileResult::applyRelocs(unsigned char* block1, ULONG blocksize1, void* o
         {
-#if defined(TARGET_X86)
-            case IMAGE_REL_BASED_HIGHLOW:
+            if (relocType == IMAGE_REL_BASED_HIGHLOW)
             {
                 DWORDLONG fixupLocation = tmp.location;
 
                 size_t address = section_begin + (size_t)fixupLocation - (size_t)originalAddr;
                 if ((section_begin <= address) && (address < section_end)) // A reloc for our section?
                 {
                     LogDebug("  fixupLoc-%016llX (@%p) : %08X => %08X", fixupLocation, address, *(DWORD*)address,
-                             (DWORD)tmp.target);
+                        (DWORD)tmp.target);
                     *(DWORD*)address = (DWORD)tmp.target;
                 }
+                wasRelocHandled = true;
             }
-            break;
-#endif // TARGET_X86
+        }
 
-#if defined(TARGET_X86) || defined(TARGET_AMD64) || defined(TARGET_ARM64) || defined(TARGET_ARM)
-            case IMAGE_REL_BASED_REL32:
+        if (targetArch == SPMI_TARGET_ARCHITECTURE_ARM)
+        {
+            DWORDLONG fixupLocation = tmp.location;
+            DWORDLONG address       = section_begin + (size_t)fixupLocation - (size_t)originalAddr;
+
+            switch (relocType)
             {
-                DWORDLONG target        = tmp.target + tmp.addlDelta;
-                DWORDLONG fixupLocation = tmp.location + tmp.slotNum;
-                DWORDLONG baseAddr      = fixupLocation + sizeof(INT32);
-                INT64     delta         = (INT64)((BYTE*)target - baseAddr);
+                case IMAGE_REL_BASED_THUMB_MOV32:
+                case IMAGE_REL_BASED_REL_THUMB_MOV32_PCREL:
+                {
+                    INT32 delta  = (INT32)(tmp.target - fixupLocation);
+                    if ((section_begin <= address) && (address < section_end)) // A reloc for our section?
+                    {
+                        PutThumb2Mov32((UINT16*)address, (UINT32)delta);
+                    }
+                    wasRelocHandled = true;
+                }
+                break;
 
-#if defined(TARGET_AMD64) || defined(TARGET_ARM64)
-                if (delta != (INT64)(int)delta)
+                case IMAGE_REL_BASED_THUMB_BRANCH24:
                 {
-                    // This isn't going to fit in a signed 32-bit address. Use something that will fit,
-                    // since we assume that original compilation fit fine. This is only an issue for
-                    // 32-bit offsets on 64-bit targets.
-                    target         = (DWORDLONG)originalAddr + (DWORDLONG)blocksize1;
-                    INT64 newdelta = (INT64)((BYTE*)target - baseAddr);
+                    INT32 delta = (INT32)(tmp.target - fixupLocation);
+                    if ((section_begin <= address) && (address < section_end)) // A reloc for our section?
+                    {
+                        if (!FitsInThumb2BlRel24(delta))
+                        {
```

Comment:
I was not sure what to do here. IIRC, if such overflow occurs in real product during crossgen the JIT would be required to re-compile the method without using this type of relocation.

Reply:
@BruceForstall Any opinion if this is desired behavior? What would/should happen during spmi collect when such relocation overflow occurs? Would the corresponding method context be included in the final collection?

Reply:
I'm not sure what would happen during spmi collection if it failed. It's supposed to throw and retry, but the retry logic for relocs in crossgen1 seems to be dependent on m_fNGenLastRetry, but I can't find how that ever gets set to true (because it's dependent on fNgenLastRetry, which is never set to true).

It looks like crossgen2 always returns IMAGE_REL_BASED_THUMB_BRANCH24 with no provision for overflow.

I would guess spmi would only record the retry compilation, but it's not clear that ever happens.